### PR TITLE
Http2 v12

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/1947

DRAFT for discussion
Not to be merged

Describe changes:
- Adds basic HTTP2 parser

Still more work to be done :
 + more commit squashing
 + connection upgrade from HTTP1
 + rules file for applayer events
 + documentation
 + bug fix for file data detection
 
Modifies #5095 with 
- rebasing
- rust closes files even on 0 length chunk (as is the case in my sample pcap)
- adding unsafe to functions for clippy
- adds incomplete function to avoid DOS by buffering too much
- handle transactions the right way
- use jsonbuilder
- various fixes

### CODE REVIEW

File: rust/src/http2/files.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use crate::core::*;
+use crate::filecontainer::*;
+
+/// Wrapper around Suricata's internal file container logic.
+#[derive(Debug)]
+//TODOask one generic structure for all parsers
```
Comment (catenacyber) [inline rust/src/http2/files.rs:23]: Should we make a unique structure for SMB, NFS, HTTP2 ?..

File: rust/src/http2/http2.rs [hunk 2]
```diff
@@ -0,0 +1,1175 @@
...
+            _ => {
+                return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                    reason: HTTP2FrameUnhandledReason::UnknownType,
+                });
+            }
+        }
+    }
+
+    fn stream_data(&mut self, dir: u8, input: &[u8], over: bool, txid: u64) {
+        //TODOask use streaming buffer directly
```
Comment (catenacyber) [inline rust/src/http2/http2.rs:635]: Should we consider every content data a file on HTTP2 ?

On HTTP1, we do it only for responses from the server, not data posted by the client

File: rust/src/http2/http2.rs [hunk 3]
```diff
@@ -0,0 +1,1175 @@
...
+            }
+            state.progress = HTTP2ConnectionState::Http2StateMagicDone;
+        } else {
+            //still more buffer
+            return AppLayerResult::incomplete(0 as u32, HTTP2_MAGIC_LEN as u32);
+        }
+    }
+
+    state.files.flags_ts = unsafe { FileFlowToFlags(flow, STREAM_TOSERVER) };
+    //TODOask use FILE_USE_DETECT ?
```
Comment (catenacyber) [inline rust/src/http2/http2.rs:956]: What is `FILE_USE_DETECT` for ?

Generally speaking, I am not sure I get what should be done to handle files the right way
Help welcome @victorjulien

File: rust/src/http2/http2.rs [hunk 4]
```diff
@@ -0,0 +1,1175 @@
...
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_getfiles(
+    state: *mut std::os::raw::c_void, direction: u8,
+) -> *mut FileContainer {
+    let state = cast_pointer!(state, HTTP2State);
+    //TODO8 understand inversion ?
```
Comment (catenacyber) [inline rust/src/http2/http2.rs:1116]: It seems `PrefilterTxFiledata` calls `AppLayerParserGetFiles` with the wrong direction

So, I guess there is an inversion somewhere...

File: rust/src/http2/parser.rs [hunk 5]
```diff
@@ -0,0 +1,1145 @@
...
+            "SETTINGS_MAX_CONCURRENT_STREAMS" => Ok(HTTP2SettingsId::SETTINGSMAXCONCURRENTSTREAMS),
+            "SETTINGS_INITIAL_WINDOW_SIZE" => Ok(HTTP2SettingsId::SETTINGSINITIALWINDOWSIZE),
+            "SETTINGS_MAX_FRAME_SIZE" => Ok(HTTP2SettingsId::SETTINGSMAXFRAMESIZE),
+            "SETTINGS_MAX_HEADER_LIST_SIZE" => Ok(HTTP2SettingsId::SETTINGSMAXHEADERLISTSIZE),
+            _ => Err(format!("'{}' is not a valid value for HTTP2SettingsId", s)),
+        }
+    }
+}
+
+//TODOask move elsewhere generic
```
Comment (catenacyber) [inline rust/src/http2/parser.rs:696]: Any ideas where I could put this Rust implementation of generic detection versus unsigned integers ?
(and replace detect-engine-uint.h one day)

File: src/detect-file-data.c [hunk 6]
```diff
@@ -103,6 +103,13 @@ void DetectFiledataRegister(void)
     DetectAppLayerMpmRegister2("file_data", SIG_FLAG_TOCLIENT, 2,
             PrefilterMpmFiledataRegister, NULL,
             ALPROTO_SMB, 0);
+    //TODO8 check min progress
```
Comment (catenacyber) [inline src/detect-file-data.c:106]: Help wanted here as well to understand how to handle files :-)

**Comment by victorjulien** [comment]
`suricata.yaml.in` should also get a `http2` config block I think

File: rust/src/http2/http2.rs (hunk 2 above)
Comment (victorjulien) [reply rust/src/http2/http2.rs:635]: The HTTP1 handling is different for historical reasons. I think we can consider everything a file in http2, but we need to be mindful of things like compression (gzip etc), chunking, multi-part. Not sure if any of these apply to http2, but for http1 consider the normalized data to be the file.

File: rust/src/http2/http2.rs [hunk 7]
```diff
@@ -0,0 +1,1175 @@
...
+                        head.ftype,
+                        &rem[..hlsafe],
+                        complete,
+                        head.flags,
+                        STREAM_TOSERVER,
+                    );
+
+                    let tx = self.find_or_create_tx(&head, &txdata, STREAM_TOSERVER);
+                    tx.handle_frame(&head, &txdata, STREAM_TOSERVER);
+                    let over = head.flags & parser::HTTP2_FLAG_HEADER_END_STREAM != 0;
```
Comment (victorjulien) [inline rust/src/http2/http2.rs:723]: maybe name this something like `eof` (or `eos` ;) )

File: rust/src/http2/http2.rs [hunk 8]
```diff
@@ -0,0 +1,1175 @@
...
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    //we may have consumed data from previous records
+                    if input.len() < HTTP2_FRAME_HEADER_LEN {
+                        return AppLayerResult::incomplete(
+                            (il - input.len()) as u32,
+                            HTTP2_FRAME_HEADER_LEN as u32,
+                        );
+                    } else {
+                        panic!("HTTP2 invalid length frame header");
```
Comment (victorjulien) [inline rust/src/http2/http2.rs:743]: can this only be the result of a bug or do we need more graceful handling here? Probably best to do a debug-validate bug with a return `AppLayerResult::err()`.

File: rust/src/http2/http2.rs [hunk 9]
```diff
@@ -0,0 +1,1175 @@
...
+
+                    //get a safe length for the buffer
+                    let (hlsafe, complete) = if rem.len() < hl {
+                        (rem.len(), false)
+                    } else {
+                        (hl, true)
+                    };
+
+                    if head.length == 0 && head.ftype == parser::HTTP2FrameType::SETTINGS as u8 {
+                        input = &rem[hlsafe..];
```
Comment (victorjulien) [inline rust/src/http2/http2.rs:802]: I see no explicit check for hlsafe (typo? hlsave?) to be > 0. Concern is potential for endless loops

File: rust/src/http2/http2.rs [hunk 10]
```diff
@@ -0,0 +1,1175 @@
...
+
+#[no_mangle]
+pub extern "C" fn rs_http2_parse_ts(
+    flow: *const Flow, state: *mut std::os::raw::c_void, _pstate: *mut std::os::raw::c_void,
+    input: *const u8, input_len: u32, _data: *const std::os::raw::c_void, _flags: u8,
+) -> AppLayerResult {
+    let state = cast_pointer!(state, HTTP2State);
+    let mut buf = build_slice!(input, input_len as usize);
+
+    if state.progress < HTTP2ConnectionState::Http2StateMagicDone {
```
Comment (victorjulien) [inline rust/src/http2/http2.rs:933]: think this should be moved to a non-unsafe function. The FFI funcs should be a simple as possible: convert unsafe input into rust types, call safe rust funcs

File: rust/src/http2/http2.rs (hunk 3 above)
Comment (victorjulien) [reply rust/src/http2/http2.rs:956]: I should revisit `FILE_USE_DETECT`. It is used to tell the API that the inspection tracker in the `File` struct should be considered when running file garbage collection.

**Comment by victorjulien** [comment]
As the first next step, can you start adding SV tests? This will aid review/testing as well.

File: rust/src/http2/files.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/http2/files.rs:23]: https://redmine.openinfosecfoundation.org/issues/3825 to track this later

File: rust/src/http2/http2.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:956]: ok, thanks

File: rust/src/http2/http2.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:635]: ok, so far, so good

File: src/detect-file-data.c (hunk 6 above)
Comment (catenacyber) [reply src/detect-file-data.c:106]: Ok, so I need to define a state different than what is done in the RFC ;-)

File: src/app-layer-http2.c [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "util-unittest.h"
+
+#include "app-layer-detect-proto.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-http2.h"
+#include "rust.h"
+
+static int HTTP2RegisterPatternsForProtocolDetection(void)
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/app-layer-http2.c:38]: Nope

File: src/detect-engine-uint.c [hunk 12]
```diff
@@ -218,13 +218,175 @@ PrefilterPacketU32Compare(PrefilterPacketHeaderValue v, void *smctx)
...
+
+    if (strlen(arg2) > 0) {
+        /*set the values*/
+        switch(arg2[0]) {
+            case '<':
+            case '>':
+                if (strlen(arg3) == 0)
+                    return NULL;
+
+                if (ByteExtractStringUint8(&u8da.arg1, 10, strlen(arg3), arg3) < 0) {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-engine-uint.c:322]: ok, this helps not testing for empty string, right ?

File: src/output-json-http2.c [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+#include "util-logopenfile.h"
+#include "util-crypt.h"
+
+#include "output-json.h"
+#include "output-json-http2.h"
+#include "rust.h"
+
+#define MODULE_NAME "LogHttp2Log"
+
+//TODOask Cannot this be generic ? and most of the file as well
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/output-json-http2.c:56]: ok, leaving it this way

File: rust/src/http2/http2.rs [hunk 14]
```diff
@@ -0,0 +1,1175 @@
...
+    pub files: HTTP2Files,
+}
+
+impl HTTP2State {
+    pub fn new() -> Self {
+        Self {
+            tx_id: 0,
+            request_frame_size: 0,
+            response_frame_size: 0,
+            dynamic_headers_ts: Vec::with_capacity(255 - parser::HTTP2_STATIC_HEADERS_NUMBER),
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/http2/http2.rs:274]: ok

File: rust/src/http2/http2.rs [hunk 15]
```diff
@@ -0,0 +1,1175 @@
...
+                return self.transactions.len() - i;
+            }
+        }
+        return 0;
+    }
+
+    fn find_child_stream_id(&mut self, sid: u32) -> u32 {
+        for i in 0..self.transactions.len() {
+            //reverse order should be faster
+            if sid == self.transactions[self.transactions.len() - 1 - i].stream_id {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/http2/http2.rs:343]: ok

File: rust/src/http2/http2.rs [hunk 16]
```diff
@@ -0,0 +1,1175 @@
...
+    ) -> &mut HTTP2Transaction {
+        if header.stream_id == 0 {
+            //special transaction with only one frame
+            //as it affects the global connection, there is no end to it
+            let mut tx = HTTP2Transaction::new();
+            self.tx_id += 1;
+            tx.tx_id = self.tx_id;
+            tx.state = HTTP2TransactionState::HTTP2StateGlobal;
+            self.transactions.push(tx);
+            let pos = self.transactions.len() - 1;
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/http2/http2.rs:364]: ok

File: rust/src/http2/http2.rs [hunk 17]
```diff
@@ -0,0 +1,1175 @@
...
+                return sid;
+            }
+        }
+        return sid;
+    }
+
+    fn find_or_create_tx(
+        &mut self, header: &parser::HTTP2FrameHeader, data: &HTTP2FrameTypeData, dir: u8,
+    ) -> &mut HTTP2Transaction {
+        if header.stream_id == 0 {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/http2/http2.rs:356]: ok

File: rust/src/http2/http2.rs [hunk 18]
```diff
@@ -0,0 +1,1175 @@
...
+        &mut self, ftype: u8, input: &[u8], complete: bool, hflags: u8, dir: u8,
+    ) -> HTTP2FrameTypeData {
+        match num::FromPrimitive::from_u8(ftype) {
+            Some(parser::HTTP2FrameType::GOAWAY) => {
+                if input.len() < HTTP2_FRAME_GOAWAY_LEN {
+                    self.set_event(HTTP2Event::InvalidFrameLength);
+                    return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                        reason: HTTP2FrameUnhandledReason::Incomplete,
+                    });
+                } else {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/http2/http2.rs:405]: ok

File: rust/src/http2/http2.rs [hunk 19]
```diff
@@ -0,0 +1,1175 @@
...
+                    Ok((hrem, hs)) => {
+                        for i in 0..hs.blocks.len() {
+                            if hs.blocks[i].error
+                                >= parser::HTTP2HeaderDecodeStatus::HTTP2HeaderDecodeError
+                            {
+                                self.set_event(HTTP2Event::InvalidHeader);
+                            }
+                        }
+                        if hrem.len() > 0 {
+                            SCLogNotice!("Remaining data for HTTP2 headers");
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/http2/http2.rs:598]: indeed

File: rust/src/http2/http2.rs (hunk 7 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:723]: of for `EOS`

File: rust/src/http2/http2.rs (hunk 8 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:743]: This is meant to be unreachable code.
But this should be caught by the TCP checks about `AppLayerResult::incomplete` 
So I remove it

File: rust/src/http2/http2.rs (hunk 9 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:802]: Even if `hlsafe` = 0 (no typo, it is meant to be safe), there should be no buffer overflow as `rem` should start after `input` (the 9 bytes of the HTTP2 frame header)

**Pull request closed without merging**